microprofile/messaging/core/src/test/java/io/helidon/microprofile/messaging/HookTest.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2020 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.helidon.microprofile.messaging;
+
+import org.eclipse.microprofile.reactive.messaging.Message;
+import org.eclipse.microprofile.reactive.messaging.spi.Connector;
+import org.junit.jupiter.api.Test;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+import javax.annotation.Priority;
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.context.Initialized;
+import javax.enterprise.event.Observes;
+import javax.enterprise.inject.spi.BeanManager;
+import javax.enterprise.inject.spi.ProcessManagedBean;
+import javax.inject.Inject;
+
+import io.helidon.microprofile.config.ConfigCdiExtension;
+import io.helidon.microprofile.messaging.hook.BasicHookBean;
+import io.helidon.microprofile.tests.junit5.AddBean;
+import io.helidon.microprofile.tests.junit5.AddExtension;
+import io.helidon.microprofile.tests.junit5.DisableDiscovery;
+import io.helidon.microprofile.tests.junit5.HelidonTest;
+
+import static javax.interceptor.Interceptor.Priority.PLATFORM_AFTER;
+
+@HelidonTest
+@DisableDiscovery
+@AddBean(BasicHookBean.class)
+@AddExtension(ConfigCdiExtension.class)
+@AddExtension(MessagingCdiExtension.class)
+public class HookTest {
+
+    @Inject
+    MessagingCdiExtension messagingCdiExtension;
+
+    AtomicLong idSeq = new AtomicLong();
+
+    private void makeConnections(@Observes @Priority(PLATFORM_AFTER + 80) @Initialized(ApplicationScoped.class) Object event,
+                                 BeanManager beanManager) {
+        System.out.println("registering");
+        messagingCdiExtension.registerMessagingMethodInvocationHook((method, message) -> {
+            MessageContext.lookup(message).put("txId", idSeq.incrementAndGet());
+            Long txId = (Long) MessageContext.lookup(message).get("txId");
+            System.out.println("TxId:" + txId + " Invoking method " + method.getName());
+        }, (method, o) -> {
+            System.out.println("output "+o);
+            Long txId = (Long) MessageContext.lookup((Message<?>) o).get("txId");
+            System.out.println("TxId:" + txId + " Invoked method " + method.getName());
+        });
+    }
+
```

Comment:
@paulparkinson Idea is that I am trying to keep context hashmap between channels with Messages itself as keys(so context can be discarded by GC when message is consumed and not referenced any more). It is possible only for some signatures like in the example:
```java 
    @Outgoing("channel1")
    public PublisherBuilder<Message<String>> publish() {
        return ReactiveStreams.of("test1", "test2", "test3")
                .map(Message::of);
    }

    @Incoming("channel1")
    @Outgoing("channel2")
    public String process(String incoming) {
        return incoming.toUpperCase();
    }

    @Incoming("channel2")
    public void consume(String incoming) {
        System.out.println(">>>" + incoming);
    }
```
Output of this POC test look like this:
```
registering
TxId:1 Invoking method process
output org.eclipse.microprofile.reactive.messaging.Message$1@4ebea12c
TxId:1 Invoked method process
TxId:2 Invoking method consume
>>>TEST1
output org.eclipse.microprofile.reactive.messaging.Message$1@44c79f32
TxId:2 Invoked method consume
TxId:3 Invoking method process
output org.eclipse.microprofile.reactive.messaging.Message$1@757d6814
TxId:3 Invoked method process
TxId:4 Invoking method consume
>>>TEST2
output org.eclipse.microprofile.reactive.messaging.Message$1@649725e3
TxId:4 Invoked method consume
TxId:5 Invoking method process
output org.eclipse.microprofile.reactive.messaging.Message$1@4c168660
TxId:5 Invoked method process
TxId:6 Invoking method consume
>>>TEST3
output org.eclipse.microprofile.reactive.messaging.Message$1@52b56a3e
TxId:6 Invoked method consume
```

Reply:
hm its not right